api/src/main/java/org/openmrs/api/context/UserContext.java

```diff
@@ -87,30 +86,42 @@ public UserContext() {
 	}
 	
 	/**
-	 * Authenticate the user to this UserContext.
-	 *
-	 * @see org.openmrs.api.context.Context#authenticate(String, String)
-	 * @param username String login name
-	 * @param password String login password
-	 * @param contextDAO ContextDAO implementation to use for authentication
-	 * @return User that has been authenticated
+	 * Authenticate user with the provided authentication scheme and credentials.
+	 * 
+	 * @param The authentication scheme to use
+	 * @param The credentials to use to authenticate
+	 * @return The authenticated client information
 	 * @throws ContextAuthenticationException
+	 * 
+	 * @since 2.3.0
 	 */
-	public User authenticate(String username, String password, ContextDAO contextDAO) throws ContextAuthenticationException {
-		log.debug("Authenticating with username: {}", username);
+	public Authenticated authenticate(AuthenticationScheme authenticationScheme, Credentials credentials)
+			throws ContextAuthenticationException {
+
+		if (log.isDebugEnabled()) {
+			log.debug("Authenticating with sheme: " + credentials.getAuthenticationScheme());
+		}
+
+		Authenticated authenticated = null;
 		try {
-			this.user = contextDAO.authenticate(username, password);
+			authenticated = authenticationScheme.authenticate(credentials);
+			this.user = authenticated.getUser();
 			notifyUserSessionListener(this.user, Event.LOGIN, Status.SUCCESS);
-		} catch(ContextAuthenticationException e) {
+		}
+		catch(ContextAuthenticationException e) {
 			User loggingInUser = new User();
-			loggingInUser.setUsername(username);
+			loggingInUser.setUsername(credentials.getClientName());
 			notifyUserSessionListener(loggingInUser, Event.LOGIN, Status.FAIL);
 			throw e;
 		}
+		
 		setUserLocation();
-		log.debug("Authenticated as: {}", this.user);
 		
-		return this.user;
+		if (log.isDebugEnabled()) {
```

Comment:
Do we need the log.isDebugEnabled() check?

Reply:
@dkayiwa those checks are all over the code base though, so I will let them in for now. I agree that it's weird, I would have assumed that a logger knows whether it supports debugging or not. Please advise.

Reply:
Are you sure there is any such in the latest version of master?

Reply:
@dkayiwa yes if you do a search, simply via GitHub like [here](https://github.com/openmrs/openmrs-core/search?q=log.isDebugEnabled%28%29&unscoped_q=log.isDebugEnabled%28%29), you will find dozens. One example on `Concept`: https://github.com/openmrs/openmrs-core/blob/master/api/src/main/java/org/openmrs/Concept.java#L451

Permalink [here](https://github.com/openmrs/openmrs-core/blob/c9996a61e8d920c1cbe5c1c8c0c4b3b3dd534fd4/api/src/main/java/org/openmrs/Concept.java#L451).

Reply:
There was a pull request which removed those in a number of classes. I just noticed that it did not fully remove them from the entire codebase. 😊